backend/src/main/kotlin/org/icpclive/cds/clics/ClicsDataSource.kt

```diff
@@ -125,26 +130,48 @@ class ClicsDataSource : ContestDataSource {
                             }
                         }
                     }
+                    is CommentaryEvent -> {
+                        analyticsEventsFlow.emit(
+                            AnalyticsCommentaryEvent(
+                                it.data.id,
+                                it.data.message,
+                                it.data.contest_time,
+                                it.data.team_ids?.firstOrNull()
+                            )
+                        )
+                        logger.info(it.data.toString())
+                    }
                     is IgnoredEvent -> {}
                 }
             }
         }
-        rawRunsFlow to contestInfoFlow
+        Triple(rawRunsFlow, contestInfoFlow, analyticsEventsFlow)
     }
 
     override suspend fun run() {
         coroutineScope {
-            val (rawRunsFlow, contestInfoFlow) = launchLoader()
-            launchICPCServices(rawRunsFlow, contestInfoFlow)
+            val (rawRunsFlow, contestInfoFlow, analyticsEventFlow) = launchLoader()
+            launchICPCServices(rawRunsFlow, contestInfoFlow, analyticsEventFlow)
         }
     }
 
     override suspend fun loadOnce() = coroutineScope {
-        val (_, contestInfoFlow) = launchLoader()
+        val (_, contestInfoFlow, analyticsEventFlow) = launchLoader()
+        val analyticsEvents = merge(contestInfoFlow, analyticsEventFlow)
+            .takeWhile { it !is ContestInfo || it.status != ContestStatus.OVER }
+            .fold(mutableListOf<AnalyticsEvent>()) { ac, it ->
```

Comment:
.filterIsInstance<AnalyticsEvent>
.mapNotNull { it as? AnalyticsEvent }

Not sure which of them is defined on flows. On the other side, I don't understand what you are trying to do. Thay can be merged in any way, not same as was in original flow, so filtering merge until some condition on one of them doesn't make much sense.